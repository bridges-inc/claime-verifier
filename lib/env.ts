const projectName = 'claime-verifier'

export enum Environments {
  PROD = 'prod',
  DEV = 'dev',
  TEST = 'test',
}
export interface EnvironmentVariables {
  allowedOrigin: string
  rootDomain: string
  hostedZoneId: string
  certificateArn: string
}

const EnvironmentVariablesSetting: {
  [key in Environments]: EnvironmentVariables
} = {
  [Environments.PROD]: {
    allowedOrigin: '',
    rootDomain: '',
    hostedZoneId: '',
    certificateArn: '',
  },
  [Environments.DEV]: {
    allowedOrigin: 'https://claime-webfront-k6p1srx99-squard.vercel.app',
    rootDomain: 'claime-dev.tk',
    hostedZoneId: 'Z08620181ARYV5PENJUEI',
    certificateArn:
      'arn:aws:acm:us-east-1:495476032358:certificate/92761d01-6e38-4ca4-8580-a945a5c379fe',
  },
  [Environments.TEST]: {
    allowedOrigin: '',
    rootDomain: 'test',
    hostedZoneId: 'test',
    certificateArn: '',
  },
}

export function valueOf(env: Environments): EnvironmentVariables {
  return EnvironmentVariablesSetting[env]
}
export const withEnvPrefix = (target: Environments, str: string) =>
  `${projectName}-${str}-${target}`

export const isProd = (target: Environments) => {
  return target === Environments.PROD
}
export const isDev = (target: Environments) => {
  return target === Environments.DEV
}
