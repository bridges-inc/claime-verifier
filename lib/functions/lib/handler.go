package lib

import (
	"os"
)

// Headers with headers
func Headers() map[string]string {
	return map[string]string{
		"Access-Control-Allow-Headers":     "*",
		"Access-Control-Allow-Methods":     "GET,POST,PUT,DELETE",
		"Access-Control-Allow-Credentials": "true",
		"Access-Control-Allow-Origin":      os.Getenv("AllowedOrigin"),
	}
}
