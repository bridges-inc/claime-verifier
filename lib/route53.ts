import {
  ARecord,
  ARecordProps,
  CnameRecord,
  HostedZone,
  IHostedZone,
  PublicHostedZone,
  RecordTarget,
} from '@aws-cdk/aws-route53'
import { Construct, Stack, StackProps } from '@aws-cdk/core'
import * as environment from './env'

export class Route53Stack extends Stack {
  public readonly hostedZone: IHostedZone
  constructor(
    scope: Construct,
    id: string,
    target: environment.Environments,
    props?: StackProps,
  ) {
    super(scope, id, props)
    const { rootDomain } = environment.valueOf(target)
    if (environment.isProd(target)) {
      return
    }
    this.hostedZone = hostedZone(this, target)
    // Vercel Domain Verification
    const vercelARecordProps: ARecordProps = {
      zone: this.hostedZone,
      recordName: `${rootDomain}`,
      target: RecordTarget.fromIpAddresses('76.76.21.21'),
    }
    new ARecord(this, 'VercelARecord', vercelARecordProps)

    new CnameRecord(this, 'VercelCNAMERecord', {
      zone: this.hostedZone,
      recordName: `www.${rootDomain}`,
      domainName: 'cname.vercel-dns.com',
    })
  }
}

const hostedZone = (
  scope: Construct,
  target: environment.Environments,
): IHostedZone => {
  return new PublicHostedZone(scope, 'HostedZone', {
    zoneName: hostedZoneName(target),
    comment: 'created by cdk',
  })
}

export const hostedZoneName = (target: environment.Environments) => {
  return environment.valueOf(target).rootDomain
}

export const hostedZoneFromId = (
  scope: Construct,
  target: environment.Environments,
) => {
  return HostedZone.fromHostedZoneAttributes(scope, `HostedZone`, {
    zoneName: hostedZoneName(target),
    hostedZoneId: environment.valueOf(target).hostedZoneId,
  })
}
