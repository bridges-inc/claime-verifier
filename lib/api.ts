import {
  DomainName,
  EndpointType,
  IResource,
  MockIntegration,
  PassthroughBehavior,
  RestApi,
  SecurityPolicy,
} from '@aws-cdk/aws-apigateway'
import * as certificatemanager from '@aws-cdk/aws-certificatemanager'
import { Code } from '@aws-cdk/aws-lambda'
import * as route53 from '@aws-cdk/aws-route53'
import * as alias from '@aws-cdk/aws-route53-targets'
import * as cdk from '@aws-cdk/core'
import { Stack } from '@aws-cdk/core'
import { resolve } from 'path'
import * as environment from './env'
import { hostedZoneFromId } from './route53'

type CordOptionsConfig = {
  methods: ('GET' | 'POST' | 'PUT' | 'DELETE')[]
  withCredentials?: true
  headers?: string[]
  origin?: string
}
export function addCorsOptions(
  apiResource: IResource,
  config: CordOptionsConfig,
) {
  const responseParameters = integrationResponseParameters(config)
  const methodResponseParameters = Object.keys(responseParameters).reduce(
    (res, key) => Object.assign(res, { [key]: true }),
    {},
  )
  apiResource.addMethod(
    'OPTIONS',
    new MockIntegration({
      integrationResponses: [{ statusCode: '200', responseParameters }],
      passthroughBehavior: PassthroughBehavior.NEVER,
      requestTemplates: { 'application/json': '{"statusCode": 200}' },
    }),
    {
      methodResponses: [
        { statusCode: '200', responseParameters: methodResponseParameters },
      ],
    },
  )
}

const integrationResponseParameters = (config: CordOptionsConfig) => {
  const { methods, withCredentials, headers, origin } = config
  const params = {
    'method.response.header.Access-Control-Allow-Headers': `'${
      headers?.join(',') || '*'
    }'`,
    'method.response.header.Access-Control-Allow-Origin': `'${origin || '*'}'`,
    'method.response.header.Access-Control-Allow-Methods': `'OPTIONS,${methods.join(
      ',',
    )}'`,
  }
  if (withCredentials) {
    Object.assign(params, {
      'method.response.header.Access-Control-Allow-Credentials': "'true'",
    })
  }
  return params
}

export const environmentVariables = (target: environment.Environments) => {
  return {
    AllowedOrigin: environment.valueOf(target).allowedOrigin,
    EnvironmentId: target,
    SubgraphEndpoint: environment.valueOf(target).subgraphEndpoint,
  }
}
const aRecord = (
  stack: Stack,
  target: environment.Environments,
  domainName: string,
  customDomain: DomainName,
) => {
  new route53.ARecord(stack, 'RestApiARecord', {
    zone: hostedZoneFromId(stack, target),
    recordName: domainName,
    target: route53.RecordTarget.fromAlias(
      new alias.ApiGatewayDomain(customDomain),
    ),
  })
}

export const code = (dirname: string) => {
  return Code.fromAsset(
    resolve(`${__dirname}/../`, 'lib', 'functions', dirname, 'bin', 'main.zip'),
  )
}
export const withCustomDomain = (
  stack: cdk.Stack,
  api: RestApi,
  domain: string,
  target: environment.Environments,
) => {
  const customDomain = api.addDomainName(
    environment.withEnvPrefix(target, 'domain'),
    customDomainProps(stack, domain, target),
  )
  aRecord(stack, target, domain, customDomain)
  return customDomain
}

const customDomainProps = (
  stack: cdk.Stack,
  domain: string,
  target: environment.Environments,
) => {
  return {
    domainName: domain,
    certificate: certificatemanager.Certificate.fromCertificateArn(
      stack,
      'Cert',
      environment.valueOf(target).certificateArn,
    ),
    securityPolicy: SecurityPolicy.TLS_1_2,
    endpointType: EndpointType.REGIONAL,
  }
}
