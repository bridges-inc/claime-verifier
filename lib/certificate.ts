import {
  DnsValidatedCertificate,
  ICertificate,
  ValidationMethod,
} from '@aws-cdk/aws-certificatemanager'
import { Construct, Stack } from '@aws-cdk/core'
import * as environment from './env'
import { hostedZoneFromId } from './route53'

export class CertificateStack extends Stack {
  public readonly certificate: ICertificate

  constructor(scope: Construct, id: string, target: environment.Environments) {
    super(scope, id)
    const { hostedZoneId } = environment.valueOf(target)
    if (environment.isProd(target)) return
    if (!hostedZoneId) throw new Error('env.hostedZoneId is requied')
    this.certificate = certificate(this, target)
  }
}

const certificate = (
  scope: Construct,
  target: environment.Environments,
): ICertificate => {
  const { rootDomain } = environment.valueOf(target)

  return new DnsValidatedCertificate(scope, 'Certificate', {
    domainName: `${rootDomain}`,
    subjectAlternativeNames: [`*.${rootDomain}`],
    hostedZone: hostedZoneFromId(scope, target),
    validationMethod: ValidationMethod.DNS,
    region: 'us-east-1',
  })
}
