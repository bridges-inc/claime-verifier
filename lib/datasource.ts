import {
  AttributeType,
  BillingMode,
  Table,
  TableProps,
} from "@aws-cdk/aws-dynamodb";
import * as cdk from "@aws-cdk/core";
import * as environment from "../lib/env";
export class DatasourceStack extends cdk.Stack {
  constructor(
    scope: cdk.Construct,
    id: string,
    target: environment.Environments,
    props?: cdk.StackProps
  ) {
    super(scope, id, props);
    const table = new Table(
      this,
      "claime-verifier-table",
      ddbProps(environment.withEnvPrefix(target, "main"))
    );
    table.addGlobalSecondaryIndex({
      indexName: "GSI-1",
      partitionKey: {
        name: "PK",
        type: AttributeType.STRING,
      },
      sortKey: {
        name: "Timestamp",
        type: AttributeType.STRING,
      },
    });
  }
}

const ddbProps = (tableName: string): TableProps => {
  return {
    tableName: tableName,
    partitionKey: {
      name: `PK`,
      type: AttributeType.STRING,
    },
    sortKey: {
      name: `SK`,
      type: AttributeType.STRING,
    },
    billingMode: BillingMode.PAY_PER_REQUEST,
    removalPolicy: cdk.RemovalPolicy.DESTROY,
    pointInTimeRecovery: true,
  };
};
