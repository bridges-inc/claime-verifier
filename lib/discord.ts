import {
  ApiKeySourceType,
  LambdaIntegration,
  RequestValidator,
  RestApi,
} from '@aws-cdk/aws-apigateway'
import { SubnetType, Vpc } from '@aws-cdk/aws-ec2'
import {
  Cluster,
  ContainerImage,
  FargateService,
  FargateTaskDefinition,
  LogDriver,
} from '@aws-cdk/aws-ecs'
import {
  Effect,
  ManagedPolicy,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from '@aws-cdk/aws-iam'
import { Code, Function, Runtime, Tracing } from '@aws-cdk/aws-lambda'
import { LogGroup } from '@aws-cdk/aws-logs'
import { Secret } from '@aws-cdk/aws-secretsmanager'
import * as cdk from '@aws-cdk/core'
import { RemovalPolicy } from '@aws-cdk/core'
import { resolve } from 'path'
import { dataSourceReadWritePolicyStatement } from './datasource'
import * as environment from './env'
import { environmentVariables } from './restapi'

export class DiscordStack extends cdk.Stack {
  constructor(
    scope: cdk.Construct,
    id: string,
    target: environment.Environments,
    props?: cdk.StackProps,
  ) {
    super(scope, id, props)
    const api = new RestApi(this, 'RestApi', {
      restApiName: environment.withEnvPrefix(target, 'restapi'),
      apiKeySourceType: ApiKeySourceType.HEADER,
    })
    const discordAPISecrets = new Secret(
      this,
      environment.withEnvPrefix(target, 'discord-api-key'),
    )
    const func = discordFunction(
      this,
      this.region,
      this.account,
      'discord',
      target,
      api,
    )
    discordAPISecrets.grantRead(func)
    const vpc = new Vpc(this, 'vpc', {
      cidr: '192.156.0.0/16',
      natGateways: 1,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'Public1',
          subnetType: SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'Public2',
          subnetType: SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'Private1',
          subnetType: SubnetType.PRIVATE,
        },
        {
          cidrMask: 24,
          name: 'Private2',
          subnetType: SubnetType.PRIVATE,
        },
      ],
    })
    const executionRole = new Role(this, 'EcsTaskExecutionRole', {
      roleName: environment.withEnvPrefix(target, 'ecs-execution-role'),
      assumedBy: new ServicePrincipal('ecs-tasks.amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AmazonECSTaskExecutionRolePolicy',
        ),
      ],
    })
    const serviceTaskRole = new Role(this, 'EcsServiceTaskRole', {
      roleName: environment.withEnvPrefix(target, 'ecs-service-task-role'),
      assumedBy: new ServicePrincipal('ecs-tasks.amazonaws.com'),
    })
    basicPolicytStatements(this.region, this.account, target).forEach((s) =>
      serviceTaskRole.addToPolicy(s),
    )
    const logGroup = new LogGroup(this, 'ServiceLogGroup', {
      logGroupName: '/aws/ecs/claime-cluster' + target,
      removalPolicy: RemovalPolicy.DESTROY,
    })
    const image = ContainerImage.fromAsset('.')
    const cpu = 256
    const mem = 1024
    const taskDef = new FargateTaskDefinition(this, 'ServiceTaskDefinition', {
      cpu: cpu,
      memoryLimitMiB: mem,
      executionRole: executionRole,
      taskRole: serviceTaskRole,
    })
    taskDef.addContainer('ContainerDef', {
      image,
      cpu: cpu,
      memoryLimitMiB: mem,
      logging: LogDriver.awsLogs({
        streamPrefix: 'claime',
        logGroup,
      }),
      environment: environmentVariables(target),
    })
    const cluster = new Cluster(this, 'claime-cluster', {
      clusterName: environment.withEnvPrefix(target, 'cluster'),
      containerInsights: true,
      vpc: vpc,
      enableFargateCapacityProviders: true,
    })
    const fargateService = new FargateService(this, 'FargateService', {
      cluster,
      vpcSubnets: vpc.selectSubnets({ subnetType: SubnetType.PRIVATE }),
      taskDefinition: taskDef,
      desiredCount: 1,
    })
  }
}

const discordFunction = (
  scope: cdk.Construct,
  region: string,
  account: string,
  resource: string,
  target: environment.Environments,
  api: RestApi,
) => {
  const requestValidator = fromDiscordRequestValidator(scope, 'validator', api)
  const func = new Function(scope, resource, {
    functionName: `${environment.withEnvPrefix(target, resource)}`,
    code: code(resource),
    handler: 'bin/main',
    timeout: cdk.Duration.minutes(1),
    runtime: Runtime.GO_1_X,
    tracing: Tracing.ACTIVE,
    environment: environmentVariables(target),
  })
  basicPolicytStatements(region, account, target).forEach((s) =>
    func.addToRolePolicy(s),
  )
  api.root.addMethod(
    'POST',
    new LambdaIntegration(func, fromDiscordIntegrationOptions()),
    discordMethodIntegration(requestValidator),
  )
  return func
}

const discordMethodIntegration = (validator: RequestValidator) => {
  return {
    requestValidator: validator,
    apiKeyRequired: false,
    methodResponses: [
      {
        statusCode: '200',
      },
      {
        statusCode: '401',
      },
    ],
  }
}

const fromDiscordRequestValidator = (
  scope: cdk.Construct,
  id: string,
  api: RestApi,
) => {
  return new RequestValidator(scope, id, {
    restApi: api,
    validateRequestBody: true,
    validateRequestParameters: true,
  })
}

const fromDiscordIntegrationOptions = () => {
  return {
    proxy: false,
    requestTemplates: {
      'application/json':
        '{\r\n\
          "timestamp": "$input.params(\'x-signature-timestamp\')",\r\n\
          "signature": "$input.params(\'x-signature-ed25519\')",\r\n\
          "jsonBody" : $input.json(\'$\')\r\n\
        }',
    },
    integrationResponses: [
      {
        statusCode: '200',
      },
      {
        statusCode: '401',
        selectionPattern: '.*[UNAUTHORIZED].*',
        responseTemplates: {
          'application/json': 'invalid request signature',
        },
      },
    ],
  }
}

export const basicPolicytStatements = (
  region: string,
  account: string,
  target: environment.Environments,
) => {
  return [
    dataSourceReadWritePolicyStatement(region, account, target),
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['ssm:Get*'],
      resources: [
        `arn:aws:ssm:${region}:${account}:parameter/claime-verifier*`,
      ],
    }),
  ]
}

const code = (dirname: string) => {
  return Code.fromAsset(
    resolve(`${__dirname}/../`, 'lib', 'functions', dirname, 'bin', 'main.zip'),
  )
}
