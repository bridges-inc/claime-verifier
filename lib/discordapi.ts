import { LambdaIntegration, RestApi } from '@aws-cdk/aws-apigateway'
import { Function, Runtime, Tracing } from '@aws-cdk/aws-lambda'
import * as cdk from '@aws-cdk/core'
import {
  addCorsOptions,
  code,
  environmentVariables,
  withCustomDomain,
} from './api'
import { basicPolicytStatements } from './discord'
import * as environment from './env'

export class DiscordApiStack extends cdk.Stack {
  constructor(
    scope: cdk.Construct,
    id: string,
    target: environment.Environments,
    props?: cdk.StackProps,
  ) {
    super(scope, id, props)
    const api = new RestApi(this, 'RestApi', {
      restApiName: environment.withEnvPrefix(target, 'discord-restapi'),
    })
    const func = discordFunction(
      this,
      this.region,
      this.account,
      'verifydiscord',
      target,
      api,
    )
    withCustomDomain(this, api, restApiDomainName(target), target)
  }
}
// GET /${eoa}?type=domain
// -> いったんbool,at,actual
// -> (ゆくゆくは検証NGだった場合に理由出したい)
const discordFunction = (
  scope: cdk.Construct,
  region: string,
  account: string,
  resource: string,
  target: environment.Environments,
  api: RestApi,
) => {
  const func = new Function(scope, resource, {
    functionName: `${environment.withEnvPrefix(target, resource)}`,
    code: code(resource),
    handler: 'bin/main',
    timeout: cdk.Duration.minutes(1),
    runtime: Runtime.GO_1_X,
    environment: environmentVariables(target),
    tracing: Tracing.ACTIVE,
  })
  basicPolicytStatements(region, account, target).forEach((s) =>
    func.addToRolePolicy(s),
  )
  const rs = api.root.addResource('verify')
  rs.addMethod('POST', new LambdaIntegration(func))
  addCorsOptions(rs, { methods: ['POST'] })
  return func
}

const restApiDomainName = (target: environment.Environments) => {
  return `discord.` + environment.valueOf(target).rootDomain
}
