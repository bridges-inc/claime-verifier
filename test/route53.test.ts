import { SynthUtils } from '@aws-cdk/assert'
import { App } from '@aws-cdk/core'
import { Environments } from '../lib/env'
import { Route53Stack } from '../lib/route53'

test('resource created', () => {
  const app = new App()
  const target = Environments.TEST
  const stack = new Route53Stack(app, 'test', target)
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
})
