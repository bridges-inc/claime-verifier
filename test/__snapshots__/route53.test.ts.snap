// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resource created 1`] = `
Object {
  "Resources": Object {
    "HostedZoneDB99F866": Object {
      "Properties": Object {
        "HostedZoneConfig": Object {
          "Comment": "created by cdk",
        },
        "Name": "test.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "VercelARecord89A76128": Object {
      "Properties": Object {
        "HostedZoneId": Object {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "test.",
        "ResourceRecords": Array [
          "76.76.21.21",
        ],
        "TTL": "1800",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "VercelCNAMERecordE66B48C8": Object {
      "Properties": Object {
        "HostedZoneId": Object {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "www.test.",
        "ResourceRecords": Array [
          "cname.vercel-dns.com",
        ],
        "TTL": "1800",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "VercelDevCNAMERecordFA0135A9": Object {
      "Properties": Object {
        "HostedZoneId": Object {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "dev.test.",
        "ResourceRecords": Array [
          "cname.vercel-dns.com",
        ],
        "TTL": "1800",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
  },
}
`;
